File: Leetcode/404.cpp
/**
 * 404. 左叶子之和
 * link：https://leetcode-cn.com/problems/sum-of-left-leaves/
 */

class Solution
{
    int sum = 0;

public:
    int sumOfLeftLeaves(TreeNode *root)
    {
        if (root != nullptr)
            dfs(root);
        return sum;
    }

    bool isLeaf(TreeNode *cur)
    {
        return cur->left == nullptr && cur->right == nullptr;
    }

    void dfs(TreeNode *cur)
    {
        if (cur->left != nullptr)
        {
            if (isLeaf(cur->left))
                sum += cur->left->val;
            else
                dfs(cur->left);
        }
        if (cur->right != nullptr)
            dfs(cur->right);
    }
};

File: Leetcode/404.py
# 404. 左叶子之和
# link：https://leetcode-cn.com/problems/sum-of-left-leaves/


class Solution:
    def sumOfLeftLeaves(self, root: TreeNode) -> int:
        def isLeaf(cur: TreeNode) -> bool:
            return not cur.left and not cur.right

        def dfs(cur: TreeNode) -> int:
            res = 0
            if cur.left:
                if isLeaf(cur.left):
                    res = cur.left.val
                else:
                    res += dfs(cur.left)

            if cur.right:
                res += dfs(cur.right)
            return res

        return dfs(root) if root else 0
